File: telemetry/telemetry-data-model/src/telemetry_parser.py
import struct
from telemetry_model import TelemetryData, create_telemetry
from serializer import FORMAT, serialize_telemetry


def parse_telemetry(packet: bytes) -> TelemetryData:
    expected_size = struct.calcsize(FORMAT)

    if len(packet) != expected_size:
        raise ValueError(
            f"Invalid packet size: expected {expected_size}, got {len(packet)}"
        )

    (
        spacecraft_id,
        timestamp,
        temperature_raw,
        voltage_raw,
        current_raw,
        sequence,
    ) = struct.unpack(FORMAT, packet)

    return TelemetryData(
        spacecraft_id=spacecraft_id,
        timestamp=timestamp,
        temperature_c=temperature_raw / 10.0,
        voltage_v=voltage_raw / 100.0,
        current_a=current_raw / 100.0,
        battery_percent=0.0,
        sequence=sequence,
    )


if __name__ == "__main__":
    telemetry = create_telemetry()
    packet = serialize_telemetry(telemetry)
    decoded = parse_telemetry(packet)

    print("Packet bytes:", packet.hex())
    print("Decoded telemetry:")
    print(decoded)

File: telemetry/telemetry-data-model/firmware/telemetry_parser.c
#include "telemetry_parser.h"

#define TELEMETRY_PACKET_SIZE 13

int parse_telemetry(
    const uint8_t *buffer,
    uint16_t length,
    TelemetryPacket *packet
)
{
    if (buffer == 0 || packet == 0)
        return 0;

    if (length != TELEMETRY_PACKET_SIZE)
        return 0;

    packet->spacecraft_id = buffer[0];

    packet->timestamp =
        ((uint32_t)buffer[1]) |
        ((uint32_t)buffer[2] << 8) |
        ((uint32_t)buffer[3] << 16) |
        ((uint32_t)buffer[4] << 24);

    packet->temperature_raw =
        (int16_t)(
            ((uint16_t)buffer[5]) |
            ((uint16_t)buffer[6] << 8)
        );

    packet->voltage_raw =
        ((uint16_t)buffer[7]) |
        ((uint16_t)buffer[8] << 8);

    packet->current_raw =
        ((uint16_t)buffer[9]) |
        ((uint16_t)buffer[10] << 8);

    packet->sequence =
        ((uint16_t)buffer[11]) |
        ((uint16_t)buffer[12] << 8);

    return 1;
}